SNIFAE Template Constructors of IValue (#34647)

terhuhf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34843 SNIFAE Template Constructors of IValue (#34647)**

Summary:
Currently, we use not_ok_to_boxing to filter Dimname that can not be
converted/constructed to IValue. The correct way should be SNIFAE the
constructor of IValue.

Test Plan:
PyTorch compiled after the code change.

All unit test passed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20494886](https://our.internmc.facebook.com/intern/diff/D20494886)

terhuhf committed:
SNIFAE Template Constructors of IValue (#34647)

Summary:
Currently, we use not_ok_to_boxing to filter Dimname that can not be
converted/constructed to IValue. The correct way should be SNIFAE the
constructor of IValue.

Test Plan:
PyTorch compiled after the code change.

All unit test passed

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terhuhf committed:
Update on "SNIFAE Template Constructors of IValue (#34647)"


Summary:
Currently, we use not_ok_to_boxing to filter Dimname that can not be
converted/constructed to IValue. The correct way should be SNIFAE the
constructor of IValue.

Test Plan:
PyTorch compiled after the code change.

All unit test passed

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -321,11 +321,23 @@ struct CAFFE2_API IValue final {
   c10::ArrayRef<IValue> toListRef() const;
 
+  // Some template constructors of IValue calls another constructor recursively.
+  // This SNIFEAs the called constructor exists.
```
nit: sfinae

ezyang commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -321,11 +321,23 @@ struct CAFFE2_API IValue final {
+  // Some template constructors of IValue calls another constructor recursively.
+  // This SNIFEAs the called constructor exists.
   template<class T>
+  using enable_if_recursive_ctor_exists =
```
`enable_if_ivalue_constructible` seems like a clearer name for this template. What do you think?

terhuhf replied:
I agree. I'll change it.

terhuhf committed:
Update on "SNIFAE Template Constructors of IValue (#34647)"


Summary:
Currently, we use not_ok_to_boxing to filter Dimname that can not be
converted/constructed to IValue. The correct way should be SNIFAE the
constructor of IValue.

Test Plan:
PyTorch compiled after the code change.

All unit test passed

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
